### ISSUE TITLE
Extremely frustrating implementation

### ISSUE THREAD

**Issue by altoid29** [cursey/safetyhook#4]
Every time that I have tried to implement SafetyHook into a new project has resulted in a **failure**. This is because adding bddisasm to the project is confusing and I'm not sure how to do properly. Everything seems to need a 1,000 different pre-installed files just for this library to work.

I've been wanting to try out SafetyHook as an alternative, "safe" hooking library, but always fail.

Can you assist me, or just create a demo project with a working implementation for MSVC?

Apologies if this comes off rude, just very frustrated; really hoping for a reply, @cursey.

**Comment by cursey** [comment]
Bddisasm isn't a very large or complex dependency. All you need from the repository is the `bddisasm.lib` library itself which can be build from two folders (`<bddisasm>/inc/**`, `<bddisasm>/bddisam/**`). You don't need to use cmake/cmkr just get `bddisasm.lib` building using the files in those folders, and get `safetyhook.lib` building using `<safetyhook>/include/**` & `<safetyhook/src/**`. Link both libraries to your final executable and it should be good to go.

**Comment by angelfor3v3r** [comment]
I thought I may as well add to this:
MSVC was the compiler used to create the project at the start, I've had success compiling on clang-cl too (in multiple projects).

I've heard complaints about how bddisasm is a big dependency, and while I agree that it is big in size, it's also very good at what it does and supports tons of instructions.

Either way, I'm working on adding Zydis and HDE as disasm libraries which you can choose from instead of bddisasm. HDE is the easiest to use since it would be included with the project (though, there are issues with it I'll talk about below).

So far I'm seeing results like this for the 32-bit test0 app:
```
clang version 15.0.1
Target: x86_64-pc-windows-msvc
```
| Library  | Size |
| ------------- | ------------- |
| bddisasm  | 511KB  |
| Zydis  | 171KB  |
| HDE | 25KB  |

Keep in mind that bddisasm/Zydis would be my recommended choices, HDE is very basic and only supports x86 with no extensions and will fail in certain cases.

I'll try my best to see if my changes would be a good addition to the library once I feel comfortable with it.

**Comment by altoid29** [comment]
Nope, still nothing. Attempting on MSVC 2019. Copied the two precompiled libraries from the repository. ( bddisasm.lib, bdshemu.lib )

For example, in InlineHook.cpp:
https://i.imgur.com/omcAyVo.png
https://i.imgur.com/jTGhC14.png

**Comment by angelfor3v3r** [comment]
Could you post the entire error log? It’s hard to tell whats actually broken from those lines. I don’t use Visual Studio but I’ll give it a try and see if it works for me.

**Comment by angelfor3v3r** [comment]
I just successfully used CMake and Visual Studio 2022 and didn't run into any issues. Here are the exact steps I took:

I created a folder at "C:/Sources/test" and opened a PowerShell terminal in it. Then I ran the commands in the following order:
```
git clone --depth 1 --branch v1.34.10 https://github.com/bitdefender/bddisasm.git
git clone --depth 1 https://github.com/cursey/safetyhook.git
cd bddisasm
cmake -G "Visual Studio 17 2022" -A x64 -B build
cmake --build build --config Release
cd ../safetyhook
cmake -G "Visual Studio 17 2022" -A x64 -B build
cmake --build build --config Release
```

Then I opened Visual Studio 2022, created an empty C++ project named "my-vs-project" and set it's location to "C:\Sources\test". Then I set the following properties:

"VC++ Directories" -> "Include Directories"
```
$(SolutionDir)../safetyhook/include
```

"VC++ Directories" -> "Library Directories"
```
$(SolutionDir)../bddisasm/build/Release
$(SolutionDir)../safetyhook/build/Release
```

"Linker" -> "Input" -> "Additional Dependencies"
```
bddisasm.lib
safetyhook.lib
```

"C/C++" -> "Language" to "ISO C++17 Standard (/std:c++17)"

I made a main.cpp with the following code:
```
#include <iostream>
#include <string>
#include <SafetyHook.hpp>

SafetyInlineHook g_hook_hello;

std::string hello(std::string name)
{
	return "Hello " + name + "!";
}

std::string hello_hook(std::string name)
{
	return g_hook_hello->original<decltype(hello_hook)>()("Foobar");
}

int32_t main()
{
	std::cout << hello("World") << '\n';

	{
		const auto factory = SafetyHookFactory::init();
		auto       builder = factory->acquire();

		g_hook_hello = builder.create_inline(hello, hello_hook);
	}

	std::cout << hello("World") << '\n';

	return 1;
}
```

Built this in Release x64 and ran the resulting file:
```
PS C:\Sources\test\my-vs-project\x64\Release> ./my-vs-project.exe
Hello World!
Hello Foobar!
```

If you want to build in x86 then you'll have to do something like this when configuring the libraries:
```
cmake -G "Visual Studio 17 2022" -A Win32 -B build
```

If this still doesn't work for you then you're doing something *very* wrong.

**Comment by angelfor3v3r** [comment]
@altoid29 The PR #13 will make using SafetyHook a lot easier. Is there any chance you could give it a try? Any testing will help, or you could wait until it's merged to test it.

**Comment by altoid29** [comment]
> @altoid29 The PR #13 will make using SafetyHook a lot easier. Is there any chance you could give it a try? Any testing will help, or you could wait until it's merged to test it.

Hey, I will most certainly give it a try once Zydis support has been natively implemented into the project, thanks for the effort on assistance.

**Comment by angelfor3v3r** [comment]
> > @altoid29 The PR #13 will make using SafetyHook a lot easier. Is there any chance you could give it a try? Any testing will help, or you could wait until it's merged to test it.
> 
> Hey, I will most certainly give it a try once Zydis support has been natively implemented into the project, thanks for the effort on assistance.

I reviewed it myself but we still need more testers. Is there any chance you could try it in your project like before to see if it works better for you? Any and all testing helps!

**Comment by NukoOoOoOoO** [comment]
I can confirm that integrating Zydis makes everything a lot easier. Here is how I did it:

1. Download the repository and drag-and-drop the files to my project
2. Change ``#include <Zydis.h>`` to ``#include "Zydis.h"`` in file ``Zydis.c`` and ``inline_hook.cpp``
3. Define macro ``ZYDIS_STATIC_BUILD`` in ``Zydis.h``
4. Link ``ntdll.lib``
5. Compile and ready to go

Using MSVC 19.33.31630 and PR #25

**Comment by angelfor3v3r** [comment]
@NukoOoOoOoO It's unfortunate that default Visual Studio (without CMake) doesn't make it much easier for you. I assume you had to go through all those changes because you didn't use CMake. I think a better route here is to just clone the repository and invoke `cmake` in it to get a lib, then just link to that and include the safetyhook include directory. Editing source code is less than ideal and probably not something safetyhook users should need to do.

**Comment by cursey** [comment]
@NukoOoOoOoO thanks for trying out #25. I'm trying to stabilize the API and I'm strongly leaning in favor of #25 as the solution going forward so I appreciate any feedback you have regarding that PR.

**Comment by NukoOoOoOoO** [comment]
So I went ahead and replaced MinHook with safetyhook for a few projects that I am/was working on, the result is weird to me. For my FFXIV project, it doesn't have any issue, but my CS:S Tas tool, it crashes when using ``original``/``call``, here is where the crash at
![image](https://user-images.githubusercontent.com/89713806/223326189-5e4eab3d-7d52-483a-b26f-af3737e71e8e.png)

**Comment by NukoOoOoOoO** [comment]
An update to my previous common:
Just found out my FFXIV project is using commit [Easy: Add an easier to use API](https://github.com/cursey/safetyhook/pull/25/commits/68a96908e61e1218e49ac780cfa2da070cded665), but my Tas tool is [Allocator: Rework allocations ](https://github.com/cursey/safetyhook/pull/25/commits/00b9d3d0447b71976a99b98fc318b365a9d45f2e)

**Comment by angelfor3v3r** [comment]
I tested the latest commit of https://github.com/cursey/safetyhook/pull/25/commits/d2215b77b72470d9e5440dee492576e571359d8f and it's still working for me on my L4D2 tool. If you don't mind, could you give me the DLL name and relative offset (RVA) to the "Addresses::Paint" function? I can make a quick test project.

**Comment by angelfor3v3r** [comment]
I made a test project and it works fine for me. I'm using clang-cl and cmkr to do it, here's some info:
```
PS> clang-cl.exe --version
clang version 15.0.1
Target: x86_64-pc-windows-msvc
Thread model: posix

] version
Build Label:           6630498   # Uniquely identifies each build
Network PatchVersion:  6630498   # Determines client and server compatibility
Protocol version:           24   # High level network protocol version
Server version:        6630498
Server AppID:           232330
Client version:        6630498
Client AppID:              240
```

And here's some code:
<details>
<summary>Code</summary>

```cpp
#include "safetyhook.hpp"
#include <cstdint>
#include <Windows.h>

enum PaintMode_t : int32_t
{
    PAINT_UIPANELS     = 1 << 0,
    PAINT_INGAMEPANELS = 1 << 1,
    PAINT_CURSOR       = 1 << 2, // software cursor, if appropriate
};

class CEngineVGui
{
public:
};

SafetyHookInline g_paint_hook;

void __thiscall CEngineVGui_Paint_hook(CEngineVGui* thisptr, PaintMode_t mode)
{
    printf("\"CEngineVGui::Paint\" mode: %i\n", mode);

    // g_paint_hook.unsafe_thiscall(thisptr, mode);
    g_paint_hook.thiscall(thisptr, mode);
}

uint32_t __stdcall init(void* args)
{
    // Try and allocate a new console for debugging purposes.
    if (!AllocConsole())
    {
        MessageBoxA(nullptr, "Unable to create a win32 console.", "Foobar", MB_ICONERROR);
        ExitProcess(0);
    }

    // Redirect output to the console.
    FILE* f;
    freopen_s(&f, "CONIN$", "r", stdin);
    freopen_s(&f, "CONOUT$", "w", stderr);
    freopen_s(&f, "CONOUT$", "w", stdout);

    const auto engine_dll        = (uintptr_t)GetModuleHandle("engine.dll");
    const auto CEngineVGui_Paint = engine_dll + 0x214280;

    g_paint_hook = safetyhook::create_inline(CEngineVGui_Paint, (uintptr_t)CEngineVGui_Paint_hook);
    if (!g_paint_hook)
    {
        printf("Failed to hook \"CEngineVGui::Paint.\"");
        return 1;
    }

    printf("Loaded\n");

    return 0;
}

BOOL WINAPI DllMain(HMODULE handle, DWORD reason, LPVOID reserved)
{
    if (reason == DLL_PROCESS_ATTACH)
    {
        DisableThreadLibraryCalls(handle);

        if (const auto thread = (HANDLE)_beginthreadex(nullptr, 0, &init, nullptr, 0, nullptr))
        {
            CloseHandle(thread);
            return 1;
        }
    }

    return 0;
}
```

</details>

**Comment by NukoOoOoOoO** [comment]
I finally got it working, it was the problem on my side.

**Comment by altoid29** [comment]
Hey, **again** I decided to give this another shot since a lot of changes and development activity seems to be going on. So I decided to try the method @NukoOoOoOoO used, but it still doesn't work. Everything besides InlineHook.cpp compiles. Also, @cursey there is various include issues in the library that I'd like to give you so you don't have to manually fix them every time you add the library to your project from Github.

Here are the issues I found, @cursey:

```**Builder.cpp**
#include "../include/safetyhook/Factory.hpp"
#include "../include/safetyhook/InlineHook.hpp"
#include "../include/safetyhook/MidHook.hpp"
#include "../include/safetyhook/ThreadFreezer.hpp"

#include "../include/safetyhook/Builder.hpp"

**Factory.hpp**
#include "Builder.hpp"

**Factory.cpp**
#include "../include/safetyhook/Builder.hpp"
#include "../include/safetyhook/Factory.hpp"
[End]

_#define NOMINMAX at start of Factory.cpp_

**InlineHook.cpp**
#include "../include/safetyhook/Factory.hpp"
#include "../include/safetyhook/InlineHook.hpp"

**MidHook.cpp**
#include "../include/safetyhook/Factory.hpp"
#include "../include/safetyhook/InlineHook.hpp"
#include "../include/safetyhook/MidHook.hpp"
[End]

**ThreadFreezer.cpp**
#include "../include/safetyhook/ThreadFreezer.hpp"
[End]```

Here's a video of me attempting to implement the library on Microsoft Visual Studio 2019 with C++ 20:
- https://streamable.com/0hdh16 (Exp. 1 day)
- https://vimeo.com/806630027

Same video, just two links for backup.

**Comment by NukoOoOoOoO** [comment]
@altoid29 I forgot to mention that I put safetyhook folder and source files together, so you don't have to manually replace the include thingy, I'm sorry about that.
![image](https://user-images.githubusercontent.com/89713806/224237786-96598740-ef09-4aa0-a56a-eb5ebc73a685.png)
And I also ended up cloning the repository and building it, so I don't have to change anything when there is an update

**Comment by altoid29** [comment]
> @altoid29 I forgot to mention that I put safetyhook folder and source files together, so you don't have to manually replace the include thingy, I'm sorry about that. ![image](https://user-images.githubusercontent.com/89713806/224237786-96598740-ef09-4aa0-a56a-eb5ebc73a685.png) And I also ended up cloning the repository and building it, so I don't have to change anything when there is an update, and here is how I did it (everything below is done in **Native Tools Command Prompt for VS**):
> 
> ```
> cd path/to/your/project
> git clone https://github.com/cursey/safetyhook
> cd safetyhook
> git fetch
> git checkout simplification // This can be ignored once the PR is merged
> cmake -A x64 -B build_x64 // Replace x64 with Win32 if you are targeting x86
> cmake --build build_x64 --config Debug // Replace Debug with Release if you want a release build
> ```
> 
> After finishing building the library, add the library include folder to your project **Include Directories** and the built static libraries folder to **Library Directories** in your project configuration properties **VC++ Directories**, then go to **Linker -> Input -> Additional Dependencies**, add the following dependencies:
> 
> ```
> safetyhook.lib
> Zydis.lib
> ntdll.lib
> ```
> 
> then it should be able to compile successfully. And I recorded a [video](https://www.youtube.com/watch?v=Z_muaTO-bgM) for it to make the instructions more clearer

Thanks for the response, I installed CMake, Git, then start Native Tools Command Prompt x64. I followed your steps and everything is smooth sailing until I get to "cmake -A x64 -B build_x64". When I do this, it says my configuration failed. Here is the error, how the hell do I fix this?

https://i.imgur.com/Kj4i4av.png

**Comment by angelfor3v3r** [comment]
Yeah, you're gonna have to use CMake for this, VS alone will not work out of the box. Make sure you're using latest CMake or a really new one for C++23

**Comment by NukoOoOoOoO** [comment]
It looks like MSVC 16.xx doesn't support C++23, maybe try using Visual Studio 2022?

**Comment by angelfor3v3r** [comment]
@altoid29 @NukoOoOoOoO There's now amalgamated builds available since 0fd71f3307f55ddbfe3308da38a005db962ac76c. You can find two Artifact zips under the Actions page for the simplification branch. Download the `safetyhook-amalgamated-zydis` zip (or the `safetyhook-amalgamated` zip if you already have Zydis in your project) and place the files in your project somewhere. Later on I believe we'll have releases so it's even easier. I'd appreciate it if you two could give these a test to see if it's even easier to add SafetyHook to your projects (preferably Visual Studio). You don't _have_ to use CMake anymore.

**Comment by cursey** [comment]
Amalgamated builds are available on the bottom of each successfully built action: https://github.com/cursey/safetyhook/actions/runs/4397304014

**Comment by altoid29** [comment]
> 

Thanks for this, do you mind also uploading the x64 & x86 .lib files and adding them as a 'Release'? I'm still unable to build them.

**Comment by altoid29** [comment]
> > > Amalgamated builds are available on the bottom of each successfully built action (where it says Artifacts): https://github.com/cursey/safetyhook/actions/runs/4397304014. Hopefully that makes this library easier for some people to use.
> > 
> > 
> > Thanks for this, do you mind also uploading the x64 & x86 .lib files and adding them as a 'Release'? I'm still unable to build them.
> 
> What errors are you getting when you try to compile?

I followed the instructions that @NukoOoOoOoO did and also used his video as a guide, but still got nowhere.

I installed CMake, Git and replaced VS2019 with VS2022.

This is the error:
![alt text](https://i.imgur.com/qixp9AT.png)

**Comment by angelfor3v3r** [comment]
@altoid29 It looks like it's defaulting to `NMake Makefiles` for you and @NukoOoOoOoO was using `Visual Studio 17 2022` for the generator (`cmake -G`). You'll have to choose `Visual Studio 17 2022` for the `-A` option to work. I personally use `Ninja` myself.

Either way, the amalgamated builds (https://github.com/cursey/safetyhook/actions/runs/4397304014) don't require you to build anything beforehand. You can just put the header/source files in your Visual Studio project, and include/build with them. We could have libraries pre-built for people but theres a lot of different compiler combinations that we would have to handle, I don't think its worth the trouble.

**Comment by angelfor3v3r** [comment]
We've been working on #25 for a while now and it's finally been merged into main. Cursey also made releases work. It's now extremely easy to get the library to work in Visual Studio without needing CMake.

The README has new installation instructions for the amalgamated builds. Read that and follow the instructions, the releases can be found here: https://github.com/cursey/safetyhook/releases

With that said, I'm going to close this issue now. There's been a lack of communication and some of my questions are left unanswered, so I'm going to assume it's not an issue or @altoid29 lost interest. If anyone has issues still, make a new issue or reopen this one with details on what's wrong.

**Comment by altoid29** [comment]
Hey, @angelfor3v3r, @cursey

Got everything working fine as of 5.12.2023.
